adr-0057: Why we don't use cloud-managed kubernetes services (#1001)
elastisys/welkin: docs/adr/0057-why-we-do-not-use-cloud-managed-kubernetes-services.md

> # Why we don't use cloud-managed Kubernetes services

Reviewer: I would phrase this in "ADR title style". How about?

```suggestion
# Do Not Use Managed Kubernetes Services
```

- Status: Accepted
- Deciders: Product Team
- Date: 2024-09-26

## Context and Problem Statement

> Some of our Infra Provider partners offer managed Kubernetes services.

Reviewer: ```suggestion
Some of our Infrastructure Providers offer managed Kubernetes services.
```

> Customers of the self-managed service have already asked why, if they are going to manage the entire stack on a modern cloud (AWS in this case), they can't use the managed Kubernetes control plane to make things easier for them.

Reviewer: These managed Kubernetes offerings range from only managing control plane all the way to offering a fully-managed Kubernetes experience.

> We know we have a lot of good reasons for why we don't (loss of fine-grained control, not equivalent offerings means a huge test matrix, etc.), and e.g. bare metal deployments would be impossible if we always assumed a managed Kubernetes Service existing underneath.

Reviewer: Please rephrase these into decision drivers and move them to the relevant section.

> We need to provide solid reasoning toward, in particular, self-managed customers (managed service ones obviously shouldn't have to care how we do what we do).

Reviewer: I think this line can be removed, I'd like the ADR not to sound like meeting notes, but a decision record.

> Should we use cloud-managed Kubernetes services with Welkin?

Reviewer: ```suggestion
Should we use Managed Kubernetes services for Kubernetes Cluster life-cycle management within Welkin?
```

## Decision Drivers

- We want to maintain Platform security and stability.
- We want to avoid operational and development complexity.
- We want to best serve Application Developers.
- We want to make the Platform Administrator life easier.

## Considered Options

> - Yes

Reviewer: ```suggestion
- Yes, always use Managed Kubernetes services.
- Yes, use Managed Kubernetes services if they provide sufficient functionality.
```

> - No

Reviewer: ```suggestion
- No, do not use Managed Kubernetes services and always rely on Kubespray or a Cluster API provider for Kubernetes Cluster life-cycle management.
```

## Decision Outcome

> Chosen option: 2 - No, and write an ADR to capture the reasons why.

Reviewer: Please harmonize with the suggestions above. Also, please follow the original ADR template:

Chosen option: "[option 1]", because [justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force | … | comes out best (see below)].

Author: Second try
Please review when time permits
Thx

### Positive Consequences

> There are several advantages of using installer included in our open source project:

Reviewer: I think this line can be removed to make it sound more "ADR-y". :smile:

- Quality Assurance: When we release an upgraded platform version, you get a well-tested package of up-to-date versions of Kubernetes and other platform level components, which we confirmed that are compatible with each other.
- Regulatory Compliance: You get the Kubernetes cluster which is configured according to the relevant EU Regulations and Directives, Best Security Practices, Industry Specific Regulations applicable in Sweden, as documented here, e.g., we use ntp.se for time synchronization.

> - Platform Portability: You can install the very same Kubernetes Platform on any cloud infrastructure, which supports Cluster API. This could become important when your customer would require installation of your software on their on-prem infrastructure or cloud within a specific geographical location or jurisdiction.

Reviewer: ```suggestion
- Platform Portability: You can install the very same Kubernetes Platform on any Infrastructure Provider, either via a Cluster API provider or Kubespray. This means that a Welkin environment looks and feels the same, no matter if its hosted on Azure, EU cloud or on-prem. Customer Application can run as required within a specific geographical location or jurisdiction, with differences being hidden away by the platform.
```

### Negative Consequences

> - Perhaps we actually do want to use e.g. Azure's managed service (AKS) at some point? An ADR that says otherwise may add friction and make such a decision more difficult to change.

Reviewer: ```suggestion
- This means we don't leverage reduced operational burden which comes with mature Kubernetes offerings, such as Azure Kubernetes Service (AKS). However, we see this as a price worth paying for a portable platform.
```

# Do Not Use Managed Kubernetes Services

- Status: Accepted
- Deciders: Product Team
- Date: 2024-09-26

## Context and Problem Statement

Several of our Infrastructure Providers offer managed Kubernetes services. These services are designed to simplify the operational overhead of managing Kubernetes control planes. Customers of our self-managed service, especially those operating on modern cloud platforms like AWS, have expressed interest in leveraging these managed control planes to streamline their stack management.
This raises the question: should we consider integrating cloud-managed Kubernetes services into our Welkin stack?

## Decision Drivers

- We want to maintain Platform security and stability.
- We want to avoid operational and development complexity.
- We want to best serve Application Developers.
- We want to make the Platform Administrator life easier.

> - We want to maintain fine-grained control over the Kubernetes stack

Reviewer: ```suggestion
- We want to maintain fine-grained control over the Kubernetes stack.
```

> - We want to reduce the complexity of the testing matrix

Reviewer: ```suggestion
- We want to reduce the complexity of the testing matrix.
```

> - We want to keep platform portability

Reviewer: ```suggestion
- We want to keep platform portability.
```

## Considered Options

- Yes
- No, we do not use Managed Kubernetes services because we want to reduce QA complexity and keep platform portability.

## Decision Outcome

Chosen option: 2 - No, we do not use Managed Kubernetes services because we want to reduce QA complexity and keep platform portability.

> While managed Kubernetes services may seem appealing, adopting them introduces challenges that conflict with our operational and strategic goals. These include losing fine-grained control over the Kubernetes stack , dealing with non-equivalent service offerings across Infrastructure Providers (resulting in a complex and unmanageable test matrix), and the inability to support bare metal deployments.

Reviewer: ```suggestion
While managed Kubernetes services may seem appealing, adopting them introduces challenges that conflict with our operational and strategic goals.
These include:

- losing fine-grained control over the Kubernetes stack, e.g., how we configure the apiserver's audit logs;
- dealing with non-equivalent service offerings across Infrastructure Providers, resulting in a complex and unmanageable test matrix; and
- the inability to support bare metal deployments.
```

# Do Not Use Managed Kubernetes Services

- Status: Accepted
- Deciders: Product Team
- Date: 2024-09-26

## Context and Problem Statement

Several of our Infrastructure Providers offer managed Kubernetes services. These services are designed to simplify the operational overhead of managing Kubernetes control planes. Customers of our self-managed service, especially those operating on modern cloud platforms like AWS, have expressed interest in leveraging these managed control planes to streamline their stack management.
This raises the question: should we consider integrating cloud-managed Kubernetes services into our Welkin stack?

## Decision Drivers

- We want to maintain Platform security and stability.
- We want to avoid operational and development complexity.
- We want to best serve Application Developers.
- We want to make the Platform Administrator life easier.
- We want to maintain fine-grained control over the Kubernetes stack.
- We want to reduce the complexity of the testing matrix.
- We want to keep platform portability.

## Considered Options

> - Yes, always use Managed Kubernetes services.
> - Yes, use Managed Kubernetes services if they provide sufficient functionality.
> - No, we do not use Managed Kubernetes services because we want to reduce QA complexity and keep platform portability.
> 
> ## Decision Outcome
> 
> Chosen option: 2 - No, we do not use Managed Kubernetes services because we want to reduce QA complexity and keep platform portability.

Reviewer 2: To clarify the option numbers and make the number for the chosen option match.
```suggestion
1. Yes, always use Managed Kubernetes services.
2. Yes, use Managed Kubernetes services if they provide sufficient functionality.
3. No, we do not use Managed Kubernetes services because we want to reduce QA complexity and keep platform portability.

## Decision Outcome

Chosen option: 3 - No, we do not use Managed Kubernetes services because we want to reduce QA complexity and keep platform portability.
```
